torch/_prims/__init__.py

```diff
@@ -884,17 +924,33 @@ def _collapse_view_helper(
         )
         raise ValueError(msg)
 
-    length = 1
-    stride = 1
-    for idx in range(start, end):
-        if idx != (end - 1):
-            if not (strides[idx] == strides[idx + 1] * shape[idx + 1]):
-                return None, None
+    if a.ndim == 0 or (end - 1 == start):
+        return shape, strides
+
+    length = shape[end - 1]
+    stride = strides[end - 1]
+    for idx in reversed(range(start, end - 1)):
+        if shape[idx] == 0 or shape[idx + 1] == 0:
+            length = 0
+            stride = 0
+            break
+
+        if shape[idx] == 1:
+            continue
+
         length = length * shape[idx]
-        stride = stride * strides[idx]
+
+        if shape[idx + 1] != 1 and not (
+            strides[idx] == strides[idx + 1] * shape[idx + 1]
+        ):
+            return None, None
 
     new_shape = shape[:start] + (length,) + shape[end:]
-    new_strides = strides[:start] + (stride,) + shape[end:]
+    new_strides = strides[:start] + (stride,) + strides[end:]
+
+    # NOTE: when the input has no elements it's restrided as if it were contiguous
+    if a.numel() == 0:
```

Comment:
Shape-tracing-friendlier way of writing this is `assert all([i > 0 for i in x.shape])`, see #77232

Reply:
What happens if you have a 0 element tensor with "non-contiguous" dimensions that you "flatten"? Eager flatten produces view in this case, but seems like you would not, as you would return `None, None` earlier? 
```
In [17]: a=torch.randn(0,3,4,5).transpose(-2,-1)

In [18]: a.stride()
Out[18]: (60, 20, 1, 5)

In [19]: a.flatten(1,3).shape
Out[19]: torch.Size([0, 60])

In [21]: a.flatten(1,3).stride()
Out[21]: (60, 1)

In [22]: a.flatten(1,3)._is_view()
Out[22]: True
```

Reply:
My guess is we'll prefer writing `a.numel()` when we actually care about the number of elements -- but if we decided otherwise this is any easy change later.

Interesting case for flatten! Let me add it to the reference inputs now

Reply:
Added to reference inputs. You were correct, @ngimel -- fixed a bug here and a bug revealed in the prims.collapse impl, too!